Dask masked-data stats

pp-mo opened the pull request:
Contributes to #2717

pp-mo commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,45 @@ def interp_order(length):
-            point_mask_counts = da.sum(da.isnan(array), axis=axis)
+            # Build a lazy computation to compare the fraction of missing
+            # input points at each output point to the 'mdtol' threshold.
+            point_counts = da.sum(da.ones(array.shape, chunks=array.chunks),
```
Because the summation may be over one or multiple dimensions (axes).
E.G. if you have shape=(2,3,4), some alternatives might be ...
```
>>> a = np.ones((2,3,4))
>>> a
array([[[ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.]],

       [[ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.]]])
>>> np.sum(a, axis=1)
array([[ 3.,  3.,  3.,  3.],
       [ 3.,  3.,  3.,  3.]])
>>> np.sum(a, axis=-1)
array([[ 4.,  4.,  4.],
       [ 4.,  4.,  4.]])
>>> np.sum(a)
24.0
>>> np.sum(a, axis=(1,2))
array([ 12.,  12.])
>>> 
```

There is almost certainly a better way of doing that, but I thought this would be okay + at least it ensures that we treat the 'axis' parameter in the prescribed numpy manner.

pp-mo commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,45 @@ def interp_order(length):
+                da.core.elemwise(ma.getmaskarray, array),
+                axis=axis)
             masked_point_fractions = (point_mask_counts + 0.0) / point_counts
             # Note: the +0.0 forces a floating-point divide.
```
It was my implementation, just not on this iteration !

I'd rather retain some explicit indication of "using floating division on ints here", for code clarity.
Although the resulting code is "tidier", I don't think it's as clear.

A `+ 0.0` is a pretty old-fashioned approach, though, so maybe I should use 'np.float' instead, which is a more explicit statement of what is required.
Would you prefer that ?

QuLogic replied:
`/` _is_ the explicit indication of floating division; `//` is the explicit indication of integer division. Every file should have the `__future__` import already and using Python 3 semantics. This should not be surprising.

pp-mo commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,45 @@ def interp_order(length):
-        dask_result = dask_nanfunction(array, axis=axis, **kwargs)
+        # Call the statistic to get the basic result (missing-data tolerant).
+        dask_result = dask_stats_function(array, axis=axis, **kwargs)
         if mdtol is None:
```
It does always bother me to do equality testing on floats, though I suppose this is only a shortcut.

So maybe this should say `mdtol is None or mdtol >= 1.0`, 
making any mdtol > 1 equivalent to mdtol == 1.
That makes practical sense, meaning "valid if there is even one good point".
( As internally in the code, we have `boolean_mask = masked_point_fractions > mdtol` )

However, it's slightly inconsistent with the effect of mdtol < 0.0 :  
Whereas "mdtol=0" means "no missing points", "mdtol<0" will mask everything, regardless.

pelson replied:
> though I suppose this is only a shortcut

Yep.

pp-mo commented:
> Improved use of dask-mask code.

Latest changes in response to https://github.com/dask/dask/pull/2301#issuecomment-323784497
It will also be nicer to use `da.ma.masked_array', when he has added it.

pp-mo commented:
> use `da.ma.masked_array', when he has added it.

Now done.
**Also** ditched the explicit cast for floating divides, as requested.

djkirkham commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,42 @@ def interp_order(length):
-            # Note: the +0.0 forces a floating-point divide.
+            # Build a lazy computation to compare the fraction of missing
+            # input points at each output point to the 'mdtol' threshold.
+            point_counts = da.sum(da.ones(array.shape, chunks=array.chunks),
```
Does this need to be lazy?

pelson replied:
This shouldn't be filling the array at all. This is the only real blocker to this being merged IMO.

pelson replied:
How about you get the consistency without the cost of the ones...

```
>>> shape = [1201230123, 123123, 123123, 1231, 3123123]
>>> a = da.empty(shape, chunks=shape)
>>> 
>>> a.shape
(1201230123, 123123, 123123, 1231, 3123123)
>>> r = da.sum(a, axis=[1, 2])
>>> r.shape
(1201230123, 1231, 3123123)
```

pp-mo commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,42 @@ def interp_order(length):
-            # Note: the +0.0 forces a floating-point divide.
+            # Build a lazy computation to compare the fraction of missing
+            # input points at each output point to the 'mdtol' threshold.
+            point_counts = da.sum(da.ones(array.shape, chunks=array.chunks),
```
No, it used not to be but I think it's arguably better if it is.  
Especially if this was quite a large calculation, then this way could chunk it.

As explained in [previous reply to @pelson above](https://github.com/SciTools/iris/pull/2743#discussion_r134248191), there is doubtless a better way of doing this than constructing another large array + collapsing it, but making it a collapse *does* have the advantage of sharing the interpretation of the 'axis' control with the main statistic.

pelson commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,42 @@ def interp_order(length):
 
     The returned value is a new function operating on dask arrays.
-    It has the call signature "stat(data, axis=-1, mdtol=None, *kwargs)".
+    It has the call signature "stat(data, axis=-1, mdtol=None, **kwargs)".
```
In general, I encourage you to use back-ticks:

"stat(data, axis=-1, mdtol=None, **kwargs)"

vs

``stat(data, axis=-1, mdtol=None, **kwargs)``

pp-mo replied:
None of this is public, so you never get to see a rendered version anyway !

pp-mo commented:
> How about you get the consistency without the cost of the ones...

Oh, **I get it !**
Stupidly, I had not spotted that this is always just a number
 -- or as I had it previously, an array full of all the same value.

Now fixed that, and added testing for all axis possibilities = None / single / multiple.

djkirkham commented on lib/iris/analysis/__init__.py:
```diff
@@ -1419,36 +1419,51 @@ def interp_order(length):
+            # Multiply the sizes of the collapsed dimensions, to get
+            # the total number of input points at each output point.
+            point_counts = np.prod([array.shape[axis_index]
+                                    for axis_index in axis_indices])
```
How about:
```
point_counts = np.prod(array.shape) / np.prod(dask_result.shape)
```

pp-mo replied:
Dead right !! 👍 
But I get even more ...
as "np.prod(a.shape)" === "a.size", we can just use that...
See following commit,

pp-mo commented:
Fully rebased to get tests re-checking.

I will review all the comments above + see what may be outstanding ...

pp-mo commented:
One tiny fix..
I checked all the comments above + I think all existing work is done.

Tests now passing, can you please check this out @djkirkham ?

djkirkham commented:
There is still a test skip here: https://github.com/SciTools/iris/blob/a8aabe621f0ac4b75861b31e8e4d3a31101b4c84/lib/iris/tests/unit/analysis/test_STD_DEV.py#L33

Other than that it looks good.

pp-mo commented:
> still a test skip here: 

Good spot !

pp-mo commented:
Fixed !!
Please re-review @djkirkham
